testsuite/integration/manualmode/src/test/java/org/wildfly/test/manual/elytron/seccontext/DynamicSSLContextTestCase.java

```diff
@@ -0,0 +1,340 @@
+            configureDefaultSSLContext(managementClient);
+
+            reloadServer(modelControllerClient);
+            deployer.deploy(DEPLOYMENT);
+        } catch (Exception e) {
+            Assert.fail();
+        }
+    }
+
+    @Test
+    @InSequence(2)
+    @OperateOnDeployment(DEPLOYMENT)
+    public void clientRunningInsideContainerUsesDynamicSSLContextTest() throws Exception {
+        // default ssl context is configured to be dynamic in the server
+        SSLSocketFactory sf = SSLContext.getDefault().getSocketFactory();
+        // both handshakes will succeed even though these ports require different truststore and keystore
+        SSLSocket socket = (SSLSocket) sf.createSocket(SERVER_ADDRESS, PORT1);
+        socket.startHandshake();
+        socket = (SSLSocket) sf.createSocket(SERVER_ADDRESS, PORT2);
+        socket.startHandshake();
+    }
+
+    @Test
+    @RunAsClient
+    @InSequence(3)
+    public void restoreConfigAndStopContainer() throws IOException {
+        try {
+            ManagementClient managementClient = new ManagementClient(TestSuiteEnvironment.getModelControllerClient(), SERVER_ADDRESS, getManagementPort(), "remote+http");
+            restoreConfiguration(managementClient);
+            deployer.undeploy(DEPLOYMENT);
+        } catch (Exception e) {
+            Assert.fail();
+        } finally {
+            serverController.stop(CONTAINER);
+            FileUtils.deleteDirectory(KEYSTORES_DIR);
+            FileUtils.deleteDirectory(KEYSTORES2_DIR);
+        }
+    }
+
+    private static void addHTTPSListener(ManagementClient managementClient, File directory, String serverKeystoreFilename, String[] keyStore, String[] keyManager, String serverTruststoreFilename, String[] trustStore, String[] trustManager, String[] sslContext, String[] httpsListener, String[] socketBinding, int port) throws IOException {
+        addSSLContext(managementClient, directory, serverKeystoreFilename, keyStore, keyManager, serverTruststoreFilename, trustStore, trustManager, sslContext);
+        ModelNode modelNode;
+
+        modelNode = createAddOperation(createAddress(socketBinding));
+        modelNode.get("port").set(port);
+        managementClient.getControllerClient().execute(modelNode);
+
+        ModelNode mn = createAddOperation(PathAddress.parseCLIStyleAddress("/subsystem=undertow/server=default-server/https-listener=" + httpsListener[httpsListener.length - 1]).toModelNode());
+        mn.get("socket-binding").set(socketBinding[socketBinding.length - 1]);
+        mn.get("ssl-context").set(sslContext[sslContext.length - 1]);
+        mn.get("enable-http2").set(true);
+
+        managementClient.getControllerClient().execute(mn);
+    }
+
+    private static void addSSLContext(ManagementClient managementClient, File directory, String keystoreFilename, String[] keyStore, String[] keyManager, String truststoreFilename, String[] trustStore, String[] trustManager, String[] sslContext1) {
```

Comment:
There are some existing test classes that could be used to simplify the necessary setup for these tests. As an example, take a look at `ElytronSslContextInUndertowSetupTask` in `UndertowTwoWaySslTestCase`. It makes use of an `AbstractElytronSetupTask` to set up the key material and then uses the `getConfigurableElements()` method to configure the key stores, key managers, trust managers, SSL contexts, and HTTPS listener using existing test classes.

Reply:
@fjuma I remember really trying to use these helper classes, but this test uses default-ssl-context which requires reload. And these classes have their own tearDown methods but if I initiated the reload myself sooner it was having some issues. I would have to double check but I think the issue was with the client, it was either unable to connect back or disconnect properly.

Reply:
@Skyllarr In `AutomaticSelfSignedCertificateGenerationTestCase`, a reload is also needed after updating the HTTPS listener configuration. The `AbstractElytronSetupTask#setup` method does perform a reload if required after the configuration steps. If undefining the `default-ssl-context` on tear down is problematic, I wonder if you could just override the `tearDown` method.

Reply:
@fjuma I think this is a problem - Unsetting of the default-ssl-context attribute puts the process in a restart-required state instead of reload-required (see the [analysis doc](https://github.com/darranl-archive/2020-wildfly-proposals/blob/db5b3971905865b47d7cd8ba143b3383fed8ff1e/elytron/WFCORE-4144-default-ssl_context-server.adoc)). Which forces me to use manual mode for the containers. This is because in managed mode the container's lifecycle is controlled by Arquillian and I cannot stop the container. 

And in our [testsuite docs](https://docs.wildfly.org/21/Testsuite.html) is written: `Note to @ServerSetup annotation: It works as expected only on non-manual containers.`.

Reply:
@Skyllarr Ah, ok, thanks for checking that. Your approach makes sense then.